enhancements/installer/packet-ipi.md

```diff
@@ -0,0 +1,197 @@
+- Prepare the Terraform code and Packet types necessary for an IPI installer
+- Making a CI job executing the provisioning scripts to test the IPI installer
+
+### Non-Goals
+
+It is outside the scope of this enhancement to provide explanations about the
+installation of infrastructure elements that are considered as required and
+owned by the user (e.g. DNS, Load Balancer).
+
+Additional enhancement requests will be created to determine the best approach for fulfilling each need.
+
+## Proposal
+
+- Define and implement Packet types in the openshift/installer
+  - High level variables should include:
+    - API Key
+    - Project ID
+  - Boostrap node variables include:
+    - Device plan (defaulting to minimum required)
+    - Facility
+    - Virtual Networks
+  - Control plane variables rely on boostrap variables and add:
+    - Device plan (defaulting to provide a good experience)
+  - Additional clusters creation should include
+    - Device plan (defaulting to provide a good experience)
+    - Facility
+    - Virtual Networks
+    - Future: IP Reservations, Hardware Reservations, Spot Market Requests
+- Write documentation to help users use and understand the Packet installer, to include:
+  - Packet usage basics (accounts and API keys)
+  - Packet resource requirements and options
+  - OpenShift Installer options
+  - Non-Packet components that build on the experience (DNS configuration)
+- Setup the CI job to have a running test suite
+
+### Implementation Details/Notes/Constraints
+
+- RHCOS must be made available and maintained.
+  Images available for provisioning may be custom or official.
+  Official images can be hosted outside of Packet, such as ContainerLinux and RancherOS. More details area available here: <https://github.com/packethost/packet-images>
+  Custom Images can be hosted on git, as described here:
+  - <https://www.packet.com/developers/docs/servers/operating-systems/custom-images/>
+  - <https://www.packet.com/developers/changelog/custom-image-handling/>
+- Packet does not offer a managed DNS solution. Route53, CloudFlare, or RFC 2136
+  (DNS Update), and NS record based zone forwarding are considerations. Existing
+  IPI drivers like BareMetal have similar limitations.
+- LoadBalancing can be achieved with the Packet CCM, which creates and maintains
+  a MetalLB deployment within the cluster.
+- Both Public and Private IPv4 and IPv6 addresses made available for each node.
+  Private and VLAN Networks can be restricted at the project level (default) or
+  node by node. While custom ASNs and BGP routing features are supported, these
+  and other networking features will not be exposed through the IPI.
+  <https://www.packet.com/cloud/network/>
+- Openshift with run directly on RHCOS directly on baremetal without need for a
+  VM layer.
+
+### Risks and Mitigations
+
+The official Packet ClusterAPI, CCM, and CSI drivers are all relatively new and
```

Comment:
Links to each of these components would be helpful.

It would also be nice to expand on the integration plans for each of these items.

ClusterAPI -- This is alluded to in the task list earlier in the doc, but I expect we'd have to fork the CAPI repo into OpenShift and then adapt it to the OpenShift machine API.  Changes will be needed to the openshift/machine-api-operator to understand how to run this new machine controller.

CCM -- This is going to hit a limitation we have in OpenShift.  We don't yet have integration for running the external cloud controller managers.  This enhancement is related: https://github.com/openshift/enhancements/pull/423/  That enhancement discusses using the out-of-tree OpenStack CCM.  As a prerequisite, we need a new operator to manage it.  That operator can be re-used for other out-of-tree CCMs, including this one.

CSI -- I don't know how CSI drivers are integrated.  Has this been investigated yet?  Are there notes we can include on the integration process and any expected challenges?

Reply:
> Links to each of these components would be helpful.

* https://github.com/kubernetes-sigs/cluster-api-provider-packet
* https://github.com/packethost/packet-ccm
* https://github.com/packethost/csi-packet

> It would also be nice to expand on the integration plans for each of these items.

Will do. I'm not sure how far out we'll need to plan for these integrations. I suspect, based on your earlier comments, that I may need another enhancement request for each of the components unless we can take advantage of existing features (especially for the CSI and CCM).

> I expect we'd have to fork the CAPI repo into OpenShift and then adapt it to the OpenShift machine API. 

That sounds reasonable. We could try to host it at the existing repository, but a `v1beta1` tag will conflict with an eventual kubernetes-sigs spec and the likely capi-packet release tag.

This is the v1alpha1 based MachineProvider spec: 
https://github.com/kubernetes-sigs/cluster-api-provider-packet/blob/v0.1.0/pkg/apis/packetprovider/v1alpha1/packetmachineproviderspec_types.go

I imagine the OpenShift v1beta1 capi-packet will build on this, the existing v1alpha3 enhancements, and changes necessary to adopt O/S capi v1beta1.

> Changes will be needed to the openshift/machine-api-operator to understand how to run this new machine controller.

I started a branch for this, https://github.com/openshift/machine-api-operator/compare/master...displague:packet?expand=1. We'll need some coordination to get those quay image tags into existence.

>  We don't yet have integration for running the external cloud controller managers.

The value of the Packet CCM is in tracking the Packet device ids so the state of k8s nodes can reflect the state of Packet devices (deleted device is a removed node, for example). The other value is in configuring the MetalLB based loadbalancer.  If there an alternative (in-tree) baremetal CCM available that could provide LoadBalancer service with the limitations that come from not being Packet API aware?

> I don't know how CSI drivers are integrated. Has this been investigated yet?

Is that to say only flex volumes are supported?
What Openshift Baremetal storage solutions are available that don't rely on cloud managed storage?

Reply:
There is no other bare-metal CCM.  

I'm OK if this enhancement doesn't answer all of the integration questions around these components, but I'd like them to be captured as open questions either to be addressed by follow-up PRs to this doc, or future separate enhancements if the topic is deep enough to deserve its own doc.